### ISSUE TITLE
peers retained in swarm non-randomly, by IP-address numerical order

### ISSUE THREAD

**Issue by reardonia** [transmission/transmission#6895]
### What is the issue?

Noticed this when looking at .resume files diagostically.  It appears that the logic for retaining peer atoms causing the lowest-numbered (by IP address) peers to be retained over higher-address-number peers.  This is triggered whenever the swarm/torrent hits the per-torrent peer cap, typically 20.

Not clear what the preference should be (time?)

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

Tr4 mainline

**Comment by tearfur** [comment]
> It appears that the logic for retaining peer atoms causing the lowest-numbered (by IP address) peers to be retained over higher-address-number peers.

It doesn't AFAICT, what makes you think that?

**Comment by reardonia** [comment]
Gathering more data...stay tuned.

**Comment by reardonia** [comment]
Question: should we use BEP42 as a means of prioritizing peers to retain?

**Comment by tearfur** [comment]
Will get back to this later.

**Comment by tearfur** [comment]
You've made several confusing claims in this ticket. The code in the peer info pulse is pretty straightforward, it works nothing like you described.

I'll try my best to guess what is actually happening:

1. Your tracker is returning peers in IP address order, so they are also stored in IP address order in the pool.
2. These freshly returned peers intially are all considered the same by `peer_info_pulse_helpers::ComparePeerInfo`.
3. `std::partial_sort()` will only move around elements within the range `[first, middle)` if the elements in `[first, last)` are all equal.

   https://github.com/transmission/transmission/blob/ec5296c8dc15c0d51014d67ef93658e5569f354c/libtransmission/peer-mgr.cc#L2365
5. The peers with bigger IP addresses (a.k.a. elements within `[middle, last)`) are kept in-place after `std::partial_sort()`, then got erased.

Also, did you actually mean BEP-40 instead of BEP-42?

**Comment by reardonia** [comment]
> I'll try my best to guess what is actually happening:
> 
> 1. Your tracker is returning peers in IP address order, so they are also stored in IP address order in the pool.

No, both private trackers I'm testing with use random address order for peer (based on time-seeding AFAICT)

> 2. These freshly returned peers initially are all considered the same by `peer_info_pulse_helpers::ComparePeerInfo`.
> 3. `std::partial_sort()` will only move around elements within the range `[first, middle)` if the elements in `[first, last)` are all equal.
>    https://github.com/transmission/transmission/blob/ec5296c8dc15c0d51014d67ef93658e5569f354c/libtransmission/peer-mgr.cc#L2365
> 4. The peers with bigger IP addresses (a.k.a. elements within `[middle, last)`) are kept in-place after `std::partial_sort()`, then got erased.

I guess? honestly haven't tried to debug this yet, just reporting the phenomenon. It's 100% reproducible. 

> 
> Also, did you actually mean BEP-40 instead of BEP-42?

Yes.

**Comment by tearfur** [comment]
Ah, I see what's happening now: (2) - (4) of my hypothesis is correct, and the reason for the peers in the pool to be sorted is simply because of the container is a sorted map:

https://github.com/transmission/transmission/blob/24f1e15767a1cd621256cf8f32287e9b1592d6d3/libtransmission/peer-mgr.cc#L289

**Comment by Pentaphon** [comment]
@reardonia I have a feeling you and Tearfur together are going to make Transmission 4.1.x the most optimized version in this client's history thus far. 

You're really on a roll spotting these long overlooked issues. Great work.
